File: project-code/cs50x/problem-set-6/sentimental-credit/credit.py
from cs50 import get_string


def luhn_check(card_number):
    total = 0
    reverse_digits = card_number[::-1]

    # Apply Luhn's Algorithm
    for i, digit in enumerate(reverse_digits):
        n = int(digit)
        if i % 2 == 1:
            n *= 2
            if n > 9:
                n -= 9
        total += n

    return total % 10 == 0


def card_type(card_number):
    if luhn_check(card_number):
        if card_number.startswith('34') or card_number.startswith('37'):
            if len(card_number) == 15:
                return "AMEX"
        elif card_number.startswith('4'):
            if len(card_number) in [13, 16]:
                return "VISA"
        elif 51 <= int(card_number[:2]) <= 55:
            if len(card_number) == 16:
                return "MASTERCARD"

    return "INVALID"


def main():
    card_number = get_string("Number: ")
    result = card_type(card_number)
    print(result)


if __name__ == "__main__":
    main()

File: cs50x/problem-set-1/credit/credit.c
#include <cs50.h>
#include <stdio.h>
#include <string.h>

// Function prototypes
bool check_luhn(long number);
string get_card_type(long number);

int main(void)
{
    long card_number;

    // Prompt the user for the credit card number
    card_number = get_long("Number: ");

    // Validate the card number using the Luhn algorithm and determine the card type
    if (check_luhn(card_number))
    {
        string card_type = get_card_type(card_number);
        printf("%s\n", card_type);
    }
    else
    {
        printf("INVALID\n");
    }

    return 0;
}

// Function to check if a card number is valid using the Luhn algorithm
bool check_luhn(long number)
{
    int sum = 0;
    bool alternate = false;

    while (number > 0)
    {
        int digit = number % 10;
        number /= 10;

        if (alternate)
        {
            digit *= 2;
            if (digit > 9)
            {
                digit -= 9;
            }
        }

        sum += digit;
        alternate = !alternate;
    }

    return (sum % 10) == 0;
}

// Function to determine the type of card based on the number
string get_card_type(long number)
{
    int length = 0;
    long temp = number;

    while (temp > 0)
    {
        temp /= 10;
        length++;
    }

    long start_digits = number;
    while (start_digits >= 100)
    {
        start_digits /= 10;
    }

    if (length == 15 && (start_digits == 34 || start_digits == 37))
    {
        return "AMEX";
    }
    else if (length == 16 && (start_digits >= 51 && start_digits <= 55))
    {
        return "MASTERCARD";
    }
    else if ((length == 13 || length == 16) && (start_digits / 10 == 4))
    {
        return "VISA";
    }
    else
    {
        return "INVALID";
    }
}
